docs/proposals/component-registration.md

```diff
@@ -0,0 +1,406 @@
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Modify ServiceSpec.Selector to select from both Pods and ExternalServiceProviders
+3. Add a new ExternalServiceProvider Controller that handles probing ExternalServiceProviders
+  - Probe results update ExternalServiceProvider.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceProvider Endpoints
+  - Readiness from ExternalServiceProvider.Status.Conditions analogous to PodStatus.Conditions
+5. Modify the Primary Components to self-register ExternalServiceProviders (one per component instance) instead of an Endpoint
+  - Group ExternalServiceProviders (component instances) with labels (e.g. "type=component", "component-type=core")
+6. Modify the Addon Components' Pod and Service registration to use similar labels (e.g. "type=component", "component-type=addon")
+7. Register component Services (with ServiceSpec.Selector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceProviders.
+8. Deprecate `/componentsstatuses` in favor of requesting a filtered list of ExternalServiceProviders
+  - ExternalServiceProviders are external-only, but have generic conditions (liveness and readiness)
+  - Endpoints are external + internal, but only have readiness (for now)
+
+
+#### API Types
+
+The following is a draft of how the API models might change.
+
+```
+type ExternalServiceProvider struct {
+	TypeMeta
+	ObjectMeta
+	Spec ExternalServiceProviderSpec
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
+	ReadinessProbe *Probe
+}
+
+type ExternalServiceProviderPort struct {
+	Name string // optional - for mapping from ServicePort
+	Protocol Protocol // TCP, UDP
+	Port int
+}
+
+type ExternalServiceProviderStatus struct {
+	Conditions []ExternalServiceProviderCondition
+}
+
+type ExternalServiceProviderCondition struct {
+	Type ExternalServiceProviderConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+
+type ExternalServiceProviderList struct {
+	TypeMeta
+	ListMeta
+	Items []ExternalServiceProvider
+}
+```
+
+```
+type ServiceStatus struct {
+	...
+	Conditions []ServiceCondition
+}
+
+type ServiceCondition struct {
+	Type ServiceConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
```

Comment:
I wasn't planning to replace it. Or rather I gave up trying to push for it. 

My initial proposal issue and proof of concept PR had a replacement almost entirely implemented but it was rejected by the reviewers who wanted to use Service instead.  It sounds like your re-proposing what I had already implemented. But no one seemed interested in having cleaner Component and ComponentStatus API resources at the time. So I was just planning to leave it as it is and let someone else deal with it  

-Karl

> On Oct 9, 2015, at 8:30 PM, Daniel Smith <notifications@github.com> wrote:
> 
> In docs/proposals/component-registration.md:
> 
> > +```
> > +
> > +```
> > +type ServiceStatus struct {
> > +	...
> > +	Conditions []ServiceCondition
> > +}
> > +
> > +type ServiceCondition struct {
> > +	Type ServiceConditionType // Readiness, Liveness
> > +	Status ConditionStatus // True, False, Unknown
> > +	LastUpdateTime util.Time
> > +	LastTransitionTime util.Time
> > +	Reason string
> > +	Message string
> > +}
> What is the type that would be served at the replacement for /componentstatuses?
> 
> —
> Reply to this email directly or view it on GitHub.
>